staging/src/k8s.io/apiserver/pkg/admission/attributes.go

```diff
@@ -32,9 +36,10 @@ type attributesRecord struct {
 	object      runtime.Object
 	oldObject   runtime.Object
 	userInfo    user.Info
+	annotations map[string]string
 }
 
-func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info) Attributes {
+func NewAttributesRecord(object runtime.Object, oldObject runtime.Object, kind schema.GroupVersionKind, namespace, name string, resource schema.GroupVersionResource, subresource string, operation Operation, userInfo user.Info, annotations map[string]string) Attributes {
```

Comment:
aren't the initial annotations always empty? Then we can get rid of all the trivial `nil` changes above.

Reply:
Because we don't support GetAnnotations(). So we have to pass a annotation map to Admit/Validate functions. Otherwise we don't have a way to read it back.

```
annotations := make(map[string]string)
admissionAttributes := admission.NewAttributesRecord(obj, nil, scope.Kind, namespace, name, scope.Resource, scope.Subresource, admission.Create, userInfo, annotations)
err = mutatingAdmission.Admit(admissionAttributes)
audit.LogAnnotations(ae, annotations)
```

Reply:
I was thinking we have a concrete Attributes object at that point with full access, but only pass the interface to the admission plugin. No need to hide a type on custruction.

Reply:
Yes, I agree with @sttts here.